

# 作業目標:

    1. 藉由固定的 dataset, 來驗證不同loss function
    2. Dataset 的特性跟我們選用的loss function 對accracy 的影響
    
    
# 作業重點: 
    請分別選用 "MSE", "binary _crossentropy"
    查看Train/test accurancy and loss rate
    

# 導入必要的函數

In [1]:
from keras.datasets import cifar10
import numpy as np
np.random.seed(10)

Using TensorFlow backend.


# 資料準備

In [2]:
#取得Keras Dataset
(x_img_train,y_label_train),(x_img_test,y_label_test)=cifar10.load_data()

In [3]:
#確認 CIFAR10 Dataset 資料維度
print("train data:",'images:',x_img_train.shape,
      " labels:",y_label_train.shape) 
print("test  data:",'images:',x_img_test.shape ,
      " labels:",y_label_test.shape) 

train data: images: (50000, 32, 32, 3)  labels: (50000, 1)
test  data: images: (10000, 32, 32, 3)  labels: (10000, 1)


In [4]:
#資料正規化
x_img_train_normalize = x_img_train.astype('float32') / 255.0
x_img_test_normalize = x_img_test.astype('float32') / 255.0

In [5]:
#針對Label 做 ONE HOT ENCODE
from keras.utils import np_utils
y_label_train_OneHot = np_utils.to_categorical(y_label_train)
y_label_test_OneHot = np_utils.to_categorical(y_label_test)
y_label_test_OneHot.shape

(10000, 10)

# 建立模型

In [6]:
from keras.models import Sequential
from keras.layers import Dense, Dropout, Activation, Flatten
from keras.layers import Conv2D, MaxPooling2D, ZeroPadding2D

In [7]:
model = Sequential()

In [ ]:
#卷積層1

In [8]:
model.add(Conv2D(filters=32,kernel_size=(3,3),
                 input_shape=(32, 32,3), 
                 activation='relu', 
                 padding='same'))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Conv2D(filters=64, kernel_size=(3, 3), 
                 activation='relu', padding='same'))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Conv2D(filters=128, kernel_size=(3, 3), 
                 activation='relu', padding='same'))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Conv2D(filters=256, kernel_size=(3, 3), 
                 activation='relu', padding='same'))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Flatten())
model.add(Dense(512, activation='relu'))
model.add(Dense(512, activation='relu'))
model.add(Dense(10, activation='softmax'))


# model.add(Conv2D(filters=32,kernel_size=(3,3),
#                  input_shape=(32, 32,3), 
#                  activation='relu', 
#                  padding='same'))

# model.add(MaxPooling2D(pool_size=(2, 2)))

# model.add(Conv2D(filters=64, kernel_size=(3, 3), 
#                  activation='relu', padding='same'))
# model.add(MaxPooling2D(pool_size=(2, 2)))

# model.add(Flatten())
# model.add(Dense(512, activation='relu'))
# model.add(Dense(10, activation='softmax'))

In [9]:
#檢查model 的STACK
print(model.summary())

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_1 (Conv2D)            (None, 32, 32, 32)        896       
_________________________________________________________________
max_pooling2d_1 (MaxPooling2 (None, 16, 16, 32)        0         
_________________________________________________________________
conv2d_2 (Conv2D)            (None, 16, 16, 64)        18496     
_________________________________________________________________
max_pooling2d_2 (MaxPooling2 (None, 8, 8, 64)          0         
_________________________________________________________________
conv2d_3 (Conv2D)            (None, 8, 8, 128)         73856     
_________________________________________________________________
max_pooling2d_3 (MaxPooling2 (None, 4, 4, 128)         0         
_________________________________________________________________
conv2d_4 (Conv2D)            (None, 4, 4, 256)         295168    
__________

# 載入之前訓練的模型

In [ ]:
try:
    model.load_weights("SaveModel/cifarCnnModel.h5")
    print("載入模型成功!繼續訓練模型")
except :    
    print("載入模型失敗!開始訓練一個新模型")

# 訓練模型

In [10]:
'''
作業:
請分別選用 "MSE", "binary _crossentropy"
查看Train/test accurancy and loss rate
'''
functions = ["categorical_crossentropy", "MSE", "binary _crossentropy"]
 
model.compile(loss="categorical_crossentropy", optimizer='adam', metrics=['accuracy'])
model.fit(x_img_train_normalize,
          y_label_train_OneHot, 
          validation_split=0.25,
          epochs=12, batch_size=128, verbose=1)



Instructions for updating:
Use tf.where in 2.0, which has the same broadcast rule as np.where



Train on 37500 samples, validate on 12500 samples
Epoch 1/12





37500/37500 [==============================] - 14s 376us/step - loss: 1.6082 - acc: 0.4038 - val_loss: 1.2659 - val_acc: 0.5339
Epoch 2/12
37500/37500 [==============================] - 6s 162us/step - loss: 1.1440 - acc: 0.5865 - val_loss: 1.0297 - val_acc: 0.6426
Epoch 3/12
37500/37500 [==============================] - 6s 160us/step - loss: 0.9295 - acc: 0.6688 - val_loss: 0.9376 - val_acc: 0.6745
Epoch 4/12
37500/37500 [==============================] - 6s 159us/step - loss: 0.7721 - acc: 0.7249 - val_loss: 0.8406 - val_acc: 0.7107
Epoch 5/12
37500/37500 [==============================] - 6s 159us/step - loss: 0.6512 - acc: 0.7705 - val_loss: 0.8179 - val_acc: 0.7171
Epoch 6/12
37500/37500 [==============================] - 6s 159us/step - loss: 0.5367 - acc: 0.8105 - val_loss: 0.8477 - val_acc: 0.7188
Epoch 7/12
37500/

In [13]:
scores = model.evaluate(x_img_test_normalize, y_label_test_OneHot)
print()
print('accuracy',scores[1])

10000/10000 [==============================] - 1s 127us/step

accuracy 0.7226


In [14]:
model.compile(loss="MSE", optimizer='adam', metrics=['accuracy'])
model.fit(x_img_train_normalize,
          y_label_train_OneHot, 
          validation_split=0.25,
          epochs=12, batch_size=128, verbose=1)

Train on 37500 samples, validate on 12500 samples
Epoch 1/12
37500/37500 [==============================] - 7s 181us/step - loss: 0.0086 - acc: 0.9423 - val_loss: 0.0460 - val_acc: 0.7158
Epoch 2/12
37500/37500 [==============================] - 6s 158us/step - loss: 0.0090 - acc: 0.9409 - val_loss: 0.0458 - val_acc: 0.7150
Epoch 3/12
37500/37500 [==============================] - 6s 160us/step - loss: 0.0074 - acc: 0.9511 - val_loss: 0.0467 - val_acc: 0.7158
Epoch 4/12
37500/37500 [==============================] - 6s 158us/step - loss: 0.0078 - acc: 0.9479 - val_loss: 0.0461 - val_acc: 0.7184
Epoch 5/12
37500/37500 [==============================] - 6s 157us/step - loss: 0.0077 - acc: 0.9494 - val_loss: 0.0465 - val_acc: 0.7162
Epoch 6/12
37500/37500 [==============================] - 6s 157us/step - loss: 0.0074 - acc: 0.9503 - val_loss: 0.0461 - val_acc: 0.7175
Epoch 7/12
37500/37500 [==============================] - 6s 157us/step - loss: 0.0060 - acc: 0.9609 - val_loss: 0.0467 - 

In [15]:
scores = model.evaluate(x_img_test_normalize, y_label_test_OneHot)
print()
print('accuracy',scores[1])

10000/10000 [==============================] - 1s 104us/step

accuracy 0.7109


In [17]:
model.compile(loss="binary_crossentropy", optimizer='adam', metrics=['accuracy'])
model.fit(x_img_train_normalize,
          y_label_train_OneHot, 
          validation_split=0.25,
          epochs=12, batch_size=128, verbose=1)

Train on 37500 samples, validate on 12500 samples
Epoch 1/12
37500/37500 [==============================] - 7s 184us/step - loss: 0.0186 - acc: 0.9936 - val_loss: 0.2448 - val_acc: 0.9465
Epoch 2/12
37500/37500 [==============================] - 6s 161us/step - loss: 0.0137 - acc: 0.9953 - val_loss: 0.2521 - val_acc: 0.9470
Epoch 3/12
37500/37500 [==============================] - 6s 161us/step - loss: 0.0136 - acc: 0.9953 - val_loss: 0.2668 - val_acc: 0.9471
Epoch 4/12
37500/37500 [==============================] - 6s 160us/step - loss: 0.0121 - acc: 0.9958 - val_loss: 0.2604 - val_acc: 0.9467
Epoch 5/12
37500/37500 [==============================] - 6s 159us/step - loss: 0.0112 - acc: 0.9961 - val_loss: 0.2646 - val_acc: 0.9471
Epoch 6/12
37500/37500 [==============================] - 6s 160us/step - loss: 0.0111 - acc: 0.9962 - val_loss: 0.2615 - val_acc: 0.9450
Epoch 7/12
37500/37500 [==============================] - 6s 160us/step - loss: 0.0125 - acc: 0.9957 - val_loss: 0.2564 - 

In [18]:
scores = model.evaluate(x_img_test_normalize, y_label_test_OneHot)
print()
print('accuracy',scores[1])

10000/10000 [==============================] - 1s 108us/step

accuracy 0.9422499885559082


loss="categorical_crossentropy", accuracy 0.7226
loss="MSE", accuracy 0.7109
loss="binary_crossentropy", accuracy 0.9422499885559082